pkg/controller/elasticsearch/filesettings/file_settings.go

```diff
@@ -0,0 +1,164 @@
+		for name, untypedDefinition := range p.Spec.Elasticsearch.SnapshotRepositories.Data {
+			definition, ok := untypedDefinition.(map[string]interface{})
+			if !ok {
+				return fmt.Errorf("invalid type for snapshot repositories settings of Elasticsearch %s/%s", es.Namespace, es.Name)
+			}
+			repoSettings, err := mutateSnapshotRepositorySettings(definition, es.Namespace, es.Name)
+			if err != nil {
+				return err
+			}
+			p.Spec.Elasticsearch.SnapshotRepositories.Data[name] = repoSettings
+		}
+		state.SnapshotRepositories = p.Spec.Elasticsearch.SnapshotRepositories
+	}
+	// just copy other settings
+	if p.Spec.Elasticsearch.ClusterSettings != nil {
+		state.ClusterSettings = p.Spec.Elasticsearch.ClusterSettings
+	}
+	if p.Spec.Elasticsearch.SnapshotLifecyclePolicies != nil {
+		state.SLM = p.Spec.Elasticsearch.SnapshotLifecyclePolicies
+	}
+	/*if p.Spec.Elasticsearch.SecurityRoleMappings != nil {
+		state.RoleMappings = p.Spec.Elasticsearch.SecurityRoleMappings
+	}
+	if p.Spec.Elasticsearch.AutoscalingPolicies != nil {
+		state.Autoscaling = p.Spec.Elasticsearch.AutoscalingPolicies
+	}
+	if p.Spec.Elasticsearch.IndexLifecyclePolicies != nil {
+		state.IndexLifecyclePolicies = p.Spec.Elasticsearch.IndexLifecyclePolicies
+	}
+	if p.Spec.Elasticsearch.IngestPipelines != nil {
+		state.IngestPipelines = p.Spec.Elasticsearch.IngestPipelines
+	}
+	if p.Spec.Elasticsearch.IndexTemplates != nil {
+		if state.IndexTemplates == nil {
+			state.IndexTemplates = &IndexTemplates{}
+		}
+		if p.Spec.Elasticsearch.IndexTemplates.ComposableIndexTemplates != nil {
+			state.IndexTemplates.ComposableIndexTemplates = p.Spec.Elasticsearch.IndexTemplates.ComposableIndexTemplates
+		}
+		if p.Spec.Elasticsearch.IndexTemplates.ComponentTemplates != nil {
+			state.IndexTemplates.ComponentTemplates = p.Spec.Elasticsearch.IndexTemplates.ComponentTemplates
+		}
+	}*/
+	s.State = state
+	return nil
+}
+
+// mutateSnapshotRepositorySettings ensures that a snapshot repository can be used across multiple ES clusters.
+// The Elasticsearch cluster name is injected in the repository settings depending on the type of the repository:
+// - "azure", "gcs", "s3": the `base_path` property is added is set to the cluster name
+// - "fs": the cluster name is appended to the `location` property
+// - "hdfs": the cluster name is appended to the `path` property
+// - "url": nothing is done, the repository is read-only
+// - "source": nothing is done, the repository is an indirection to another repository
+func mutateSnapshotRepositorySettings(snapshotRepository map[string]interface{}, esNs string, esName string) (map[string]interface{}, error) {
+	untypedSettings := snapshotRepository["settings"]
+	if untypedSettings == nil {
+		untypedSettings = map[string]interface{}{}
+	}
+
+	suffix := fmt.Sprintf("%s-%s", esNs, esName)
+	settings, ok := untypedSettings.(map[string]interface{})
+	if !ok {
+		return nil, errors.New("invalid type for snapshot repository settings")
```

Comment:
again, would it help to know the type here when it's invalid?

Reply:
Yes, I add it.